Support for moving RSA keys to Yubikey or NFC smart card

josecastillo opened the pull request:
This pull request implements an enhancement mentioned in #875: the ability to move a secret key from OpenKeychain to a Yubikey or NFC smart card. As discussed in #1256, on-card key generation is unacceptably slow, but the ability to generate keys in OpenKeychain and move them to a smart card is still desirable. 

Subkeys in the Edit Key activity now support a "Move Subkey to Yubikey / Smart Card" action, which will check that the key is smart card compatible, and launch the NFC Operation activity to move the key to the smart card. The command automatically determines the appropriate slot on the card based on the key flags: SIGN or CERTIFY will go in the signing slot; ENCRYPT will go in the decryption slot; and AUTHENTICATE will go in the authentication slot (unless the subkey also has flags for SIGN or CERTIFY, in which case it will go in the signing slot instead). 

Two new methods in the Base NFC Activity support the keytocard operation: ```nfcPutData```, which is responsible for setting the fingerprints and key generation timestamps, and ```nfcPutKey```, which is responsible for extracting the raw secret key parameters and sending them to the card in an appropriate format. These commands require verification of the Admin PIN, which I've hard-coded to the default "12345678" for now. This makes it simpler to put keys on a new card, but a future enhancement might involve prompting for the Admin PIN as well. I've also added a command, ```nfcModifyPIN```, to change the PIN on a smart card; I didn't add UI for this, but it will probably be necessary for a user to fully customize their smart card or Yubikey with OpenKeychain. 

I also introduce one additional UI tweak in this pull request. In the key creation activity, if the user taps a completely blank smart card (one with no key fingerprints), a dialog is displayed asking the following: "Would you like to generate a smart card compatible key?" If the user says "Yes", a flag is set, and the key generation parameters are modified to generate the following: 

1. an RSA-2048 master key for signing and key certification, 
2. an RSA-2048 subkey for encryption, and
3. an RSA-2048 subkey for authentication. 

I welcome code review and any suggestions for changes or additions. There are a few areas that I have specific questions on, which I'll call out with comments on the pull request.

Valodim commented:
awesome work so far! :+1:

josecastillo commented:
Thanks for the suggestions! Input pending results ended up being ideal for this; I was able to move the keytocard process entirely into the PgpKeyOperation, which simplifies the data flow and allows for multiple keys to be moved to the card in a single operation. The edit key UI sets a boolean flag ```mMoveKeyToCard``` in each SubkeyChange to indicate that a given subkey should be moved to a card. Once the NFC activity is complete, the operation grabs the serial number from the returned data, sets ```mDummyDivert```, and completes. 

I still have the concern that I'm using the RequiredInputParcel and CryptoInputParcel to map subkey ID's to card serial numbers, when they were intended to map input hashes to output hashes, but I wonder whether that's more of an issue of how they're named than how they're used; in the end, this works to send a list of data into the NFC operation and get a map back out.

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/PgpKeyOperation.java:
```diff
@@ -1481,4 +1537,29 @@ private static boolean isDivertToCard(PGPSecretKey secretKey) {
+                algorithm != PublicKeyAlgorithmTags.RSA_SIGN &&
+                algorithm != PublicKeyAlgorithmTags.RSA_GENERAL) {
+            log.add(OperationResult.LogType.MSG_K2C_ERROR_BAD_ALGO, indent + 1);
+            return true;
```
true?

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/operations/results/OperationResult.java:
```diff
@@ -736,6 +739,11 @@ public void onAction() {
         MSG_EXPORT_LOG_EXPORT_SUCCESS (LogLevel.OK, R.string.msg_export_log_success),
+
+        // NFC keytocard
+        MSG_K2C_ERROR_BAD_ALGO(LogLevel.ERROR, R.string.edit_key_error_bad_nfc_algo),
```
these are still part of the modify operation, should be MF

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/EditKeyFragment.java:
```diff
@@ -415,15 +419,54 @@ public void handleMessage(Message message) {
+                                mSubkeysAdapter.getSecretKeyType(position);
+                        if (secretKeyType == CanonicalizedSecretKey.SecretKeyType.DIVERT_TO_CARD ||
+                            secretKeyType == CanonicalizedSecretKey.SecretKeyType.GNU_DUMMY) {
+                            Toast.makeText(EditKeyFragment.this.getActivity(),
```
use Snackbar available from our Notify class instead of Toast

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/PgpKeyOperation.java:
```diff
@@ -356,10 +361,16 @@ public PgpEditKeyResult createSecretKeyRing(SaveKeyringParcel saveParcel) {
             CryptoInputParcel cryptoInput,
-            SaveKeyringParcel saveParcel) {
+            SaveKeyringParcel saveParcel,
+            OperationLog log) {
```
log should always be passed together with indent, if it's supposed to be logged as part of another action

Valodim commented:
I think it's fine, since the data in the RequiredInputParcel is explicitly typed by RequiredInputType. Since that data always looks the same, it's easier to use an enum there rather than subclasses, especially since parcelables require a lot of boilerplate to marshall around. It would probably make sense to change mInputHashes to something like mNfcInputData at this point though.

I added some minor comments here and there but looks great otherwise. I would be very grateful if you could add unit tests to the PgpKeyOperationTest class for the mMoveKeyToCard attribute.

dschuermann commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/CreateKeyActivity.java:
```diff
@@ -116,23 +122,45 @@ protected void onNfcPerform() throws IOException {
+        } else {
+            AlertDialog.Builder builder = new AlertDialog.Builder(this);
+            builder.setTitle("Blank Smart Card / Yubikey Detected")
+                    .setMessage("Would you like to generate a smart card compatible key?")
```
These strings should be added to strings.xml
Maybe, it fits better into the wizard-like process to ask this as part of a fragment instead of a dialog, but this is not important for this pull request. I can also do this later for you.

josecastillo commented:
Fixed the issues mentioned above, and added a test case for the keytocard functionality. Note that in the PgpKeyOperationTest class, I changed the third test subkey from an ELG-1024 key to an RSA-2048 key, so that I would have a smart card compatible key for the test. This change does not affect any of the other test cases. 

I also changed one of the SubkeyChange constructors to take a moveKeyToCard boolean instead of a dummyDivert byte array, since mDummyDivert only ever gets set during the edit key operation. The constructor was no longer used anywhere except in test cases, so I changed it there. The actual functionality of changing a key to divert-to-card is still tested, in the new keytocard test case. 

Thanks for your help and let me know if there's anything else!

josecastillo commented:
I can also squash at least the first five commits to clean up the history; let me know if you'd like me to do that.

Valodim commented:
great! there's some open questions on how particular states of yubikey and secret key are handled which we should think about, for example when uploading a secret key to a non-empty yubikey, and how to connect the create-key flow to the upload one.

but the bulk is there and looks stable enough for testing, I think the rest can all be handled in issues. no need to squash, the history looks fine as it is. if dominik has no objections we can merge this.
